src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/StringBuilderAppendShouldNotTakeSubstring.cs

```diff
@@ -0,0 +1,135 @@
+
+        internal static DiagnosticDescriptor RuleReplaceOneParameter = new DiagnosticDescriptor(
+            RuleIdOneParameterId,
+            s_localizableTitle,
+            s_localizableMessageDefault,
+            DiagnosticCategory.Usage,
+            DiagnosticHelpers.DefaultDiagnosticSeverity,
+            isEnabledByDefault: DiagnosticHelpers.EnabledByDefaultIfNotBuildingVSIX,
+            description: s_localizableDescription,
+            helpLinkUri: "", // TODO (might require an addition to docs repository?
+            customTags: WellKnownDiagnosticTags.AnalyzerException);
+
+        internal static DiagnosticDescriptor RuleReplaceTwoParameter = new DiagnosticDescriptor(
+            RuleIdTwoParameterId,
+            s_localizableTitle2,
+            s_localizableMessageDefault2,
+            DiagnosticCategory.Usage,
+            DiagnosticHelpers.DefaultDiagnosticSeverity,
+            isEnabledByDefault: DiagnosticHelpers.EnabledByDefaultIfNotBuildingVSIX,
+            description: s_localizableDescription2,
+            helpLinkUri: "", // TODO (might require an addition to docs repository?
+            customTags: WellKnownDiagnosticTags.AnalyzerException);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(RuleReplaceOneParameter, RuleReplaceTwoParameter);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            context.RegisterSyntaxNodeAction(this.Action, SyntaxKind.InvocationExpression);
+        }
+
+        private void Action(SyntaxNodeAnalysisContext context)
+        {
+            if (context.Node is InvocationExpressionSyntax invocation)
+            {
+                var invokedSymbol = context.SemanticModel.GetSymbolInfo(invocation);
+                var invokedClass = invokedSymbol.Symbol.ContainingSymbol;
+                var invokedMethod = invokedSymbol.Symbol.Name;
+                var namedType = invokedClass as INamedTypeSymbol;
+
+                if (namedType.ContainingNamespace.Name == "Text" 
+                    && namedType.ContainingNamespace.ContainingNamespace.Name == "System"
+                    && namedType.ContainingNamespace.ContainingNamespace.ContainingNamespace.IsGlobalNamespace
+                    && invokedClass.Name == "StringBuilder" 
+                    && invokedMethod == "Append"
+                    && invokedSymbol.Symbol is IMethodSymbol methodSymbol)
+                {
+                    var parameters = methodSymbol.Parameters;
+
+                    if (parameters.Length == 1 && parameters[0].Type.Name == "String")
+                    {
+                        ArgumentSyntax argument = invocation.ArgumentList.Arguments.First();
+
+                        // todo: what if there's ref? argument.RefOrOutKeyword
```

Comment:
[IInvocationOperation.Arguments](http://source.roslyn.io/#Microsoft.CodeAnalysis/Operations/IInvocationOperation.cs,47) gives you arguments with [matched parameters](http://source.roslyn.io/#Microsoft.CodeAnalysis/Operations/IArgumentOperation.cs,26), and the parameter gives you the `RefKind`.

Reply:
thanks. The question here is more a semantical one according the analyzers feature set: Do I have to care about ref?
I wasn't sure if it's possible to use ref on an argument at the method call when the method doesn't declare it to be ref.
=> after checking the docs I can delete this TODO.